hub/core/compression/base.py

```diff
@@ -0,0 +1,139 @@
+        self._msgpack = numcodecs.MsgPack()
+
+    @property
+    def __name__(self):
+        raise NotImplementedError()
+
+    def encode(self, arr: np.ndarray) -> bytes:
+        """
+        Encode array with one or multiple images.
+
+        Example:
+            img = np.ones((4, 100, 100, 1))
+            encoding = codec.encode(img)
+
+        Args:
+            arr (np.ndarray): Image data to be encoded. Can contain image/s with shapes:
+                (N, M) if image has one channel and single_channel is False,
+                (N, M, 1) if image has one channel and single_channel is True,
+                (N, M, 3). In case of multiple images they should be stacked across the first axis.
+
+        Raises:
+            ValueError: If the shape length of input array is
+                less than the number of expected dimensions.
+
+        Returns:
+            Encoded dictionary of images and metadata.
+        """
+        append_one = False
+        if self.single_channel and arr.shape[-1] == 1:
+            arr = np.reshape(arr, arr.shape[:-1])
+            append_one = True
+        if not self.single_channel or append_one:
+            shape_dims = 2
+        else:
+            shape_dims = 3
+        if len(arr.shape) < shape_dims:
+            raise ValueError(
+                f"The shape length {len(arr.shape)} of the given array should "
+                f"be greater than the number of expected dimensions {shape_dims}"
+            )
+        if len(arr.shape) == shape_dims:
+            return self._msgpack.encode(
+                [
+                    {
+                        "items": self.encode_single_image(arr),
+                        "append_one": append_one,
+                        "image_shape": arr.shape,
+                    }
+                ]
+            )
+        else:
+            image_shape = arr.shape[-shape_dims:]
+            items_shape = arr.shape[:-shape_dims]
+            items = []
+            for i in np.ndindex(items_shape):
+                items.append(self.encode_single_image(arr[i]))
+            return self._msgpack.encode(
+                [
+                    {
+                        "items": items,
+                        "items_shape": items_shape,
+                        "image_shape": image_shape,
+                        "dtype": str(arr.dtype),
+                        "append_one": append_one,
+                    }
+                ]
+            )
+
+    def decode(self, buf: bytes) -> np.ndarray:
+        """
+        Decode images from buffer.
+
+        Example:
+            images_decoded = codec.decode(images_encoded)
+
+        Args:
+            buf (bytes): Encoded images and metadata.
+
+        Returns:
+            Decoded image or array with multiple images.
+        """
+        data = self._msgpack.decode(buf)[0]
+        pass_shape = True if self.__name__ == "webp" else False
+        if "items_shape" not in data:
```

Comment:
Can this if-else be simplified or broken into methods?

We have a lot of nested levels and for loops inside the nested if-elses.